abstract-integers/src/abstract_int.rs

```diff
@@ -0,0 +1,502 @@
+                }
+                big_x.into()
+            }
+
+            /// Returns 2 to the power of the argument
+            #[allow(dead_code)]
+            pub fn pow2(x: usize) -> $name {
+                BigInt::from(1u32).shl(x).into()
+            }
+
+            /// Gets the `i`-th least significant bit of this integer.
+            #[allow(dead_code)]
+            pub fn bit(self, i: usize) -> bool {
+                assert!(
+                    i < self.b.len() * 8,
+                    "the bit queried should be lower than the size of the integer representation: {} < {}",
+                    i,
+                    self.b.len() * 8
+                );
+                let bigint : BigInt = self.into();
+                let tmp: BigInt = bigint >> i;
+                (tmp & BigInt::one()).to_bytes_le().1[0] == 1
+            }
+        }
+
+        impl From<BigUint> for $name {
+            fn from(x: BigUint) -> $name {
+                Self::from(BigInt::from(x))
+            }
+        }
+        
+        impl From<BigInt> for $name {
+            fn from(x: BigInt) -> $name {
+                let max_value = Self::max();
+                assert!(x < max_value, "{} is too large for type {}!", x, stringify!($name));
+                let repr = x.to_bytes_be().1;
+                if repr.len() > ($bits + 7) / 8 {
+                    panic!("{} is too large for type {}", x, stringify!($name))
+                }
+                let mut out = [0u8; ($bits + 7) / 8];
+                let upper = out.len();
+                let lower = upper - repr.len();
+                out[lower..upper].copy_from_slice(&repr);
+                $name {
+                    b: out,
+                    sign: Sign::Plus,
+                    signed: $signed,
+                }
+            }
+        }
+
+        impl Default for $name {
+            fn default() -> $name {
+                $name {
+                    b: [0u8; ($bits + 7) / 8],
+                    sign: Sign::Plus,
+                    signed: $signed,
+                }
+            }
+        }
+
+        impl Into<BigInt> for $name {
+            fn into(self) -> BigInt {
+                BigInt::from_bytes_be(self.sign, &self.b)
+            }
+        }
+
+        impl Into<BigUint> for $name {
+            fn into(self) -> BigUint {
+                BigUint::from_bytes_be(&self.b)
+            }
+        }
+
+        impl std::fmt::Display for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                let uint: BigInt = (*self).into();
+                write!(f, "{}", uint)
+            }
+        }
+
+        impl std::fmt::Debug for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                let uint: BigInt = (*self).into();
+                write!(f, "{}", uint)
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! abstract_public {
```

Comment:
You're making here the distinction between secret and public abstract integers. Is the purpose of this distinction tagging-related or constant-time-related ?

Reply:
both, kind-off. For now it's really only tagging and not providing certain functions. But we should add an issue to eventually get mostly ct bigint.